File: acme-admin/src/main/java/com/acme/admin/controller/BillingController.java
package com.acme.admin.controller;

import com.acme.admin.dto.BillingDtos;
import com.acme.admin.service.BillingService;
import org.springframework.http.*;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/billing")
public class BillingController {

    private final BillingService svc;

    public BillingController(BillingService svc) {
        this.svc = svc;
    }

    @PostMapping("/checkout-session")
    @PreAuthorize("hasAuthority('BILLING_CREATE')")
    public ResponseEntity<BillingDtos.CheckoutSessionResponse> createCheckoutSession(
        @RequestBody BillingDtos.CreateCheckoutSessionRequest req
    ) throws Exception {
        return ResponseEntity.status(HttpStatus.CREATED).body(svc.createCheckoutSession(req));
    }
}

File: acme-admin-dotnet/Controllers/BillingController.cs
using Acme.Admin.Api.DTO;
using Acme.Admin.Api.Security;
using Acme.Admin.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Text.Json;

namespace Acme.Admin.Api.Controllers;

[ApiController]
[Route("api/billing")]
public sealed class BillingController(BillingService service, StripePaymentService stripePaymentService) : ControllerBase
{
    [HttpPost("checkout-session")]
    [Authorize(Policy = PermissionPolicies.BillingCreate)]
    public async Task<ActionResult<BillingDtos.CheckoutSessionResponse>> CreateCheckoutSession([FromBody] BillingDtos.CreateCheckoutSessionRequest req, CancellationToken ct)
    {
        var created = await service.CreateCheckoutSessionAsync(req, ct);
        return StatusCode(StatusCodes.Status201Created, created);
    }

    [HttpPost("webhook")]
    [AllowAnonymous]
    public async Task<ActionResult<string>> Webhook(
        [FromHeader(Name = "Stripe-Signature")] string? signature,
        [FromBody] JsonElement payload,
        CancellationToken ct)
    {
        return Ok(await service.SaveWebhookAsync(signature, payload.GetRawText(), ct));
    }

    [HttpPost("payment-intents")]
    [Authorize(Policy = PermissionPolicies.BillingPay)]
    public async Task<ActionResult<StripePaymentDtos.CreatePaymentIntentResponse>> CreatePaymentIntent(
        [FromBody] StripePaymentDtos.CreatePaymentIntentRequest req,
        CancellationToken ct)
    {
        return Ok(await stripePaymentService.CreateAsync(req, ct));
    }
}
